hbase-coprocessor/src/main/java/org/apache/omid/transaction/TransactionVisibilityFilter.java

```diff
@@ -0,0 +1,248 @@
+            return ReturnCode.SKIP;
+        }
+
+        if (familyDeletionCache.containsKey(Bytes.toString(CellUtil.cloneFamily(v)))
+                && familyDeletionCache.get(Bytes.toString(CellUtil.cloneFamily(v))) >= v.getTimestamp()) {
+            return ReturnCode.NEXT_COL;
+        }
+
+        if (isCellInSnapshot(v)) {
+
+            if (CellUtils.isTombstone(v)) {
+                return ReturnCode.NEXT_COL;
+            }
+
+            if (hbaseTransaction.getVisibilityLevel() == AbstractTransaction.VisibilityLevel.SNAPSHOT) {
+                return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
+            } else if (hbaseTransaction.getVisibilityLevel() == AbstractTransaction.VisibilityLevel.SNAPSHOT_ALL) {
+                if (snapshotFilter.getTSIfInTransaction(v, hbaseTransaction).isPresent()) {
+                    return runUserFilter(v, ReturnCode.INCLUDE);
+                } else {
+                    return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
+                }
+            }
+        }
+        return ReturnCode.SKIP;
+    }
+
+
+    private ReturnCode runUserFilter(Cell v, ReturnCode snapshotReturn)
+            throws IOException {
+
+        if (userFilter == null) {
+            return snapshotReturn;
+        }
+
+        ReturnCode userRes = userFilter.filterKeyValue(v);
+        switch (userRes) {
+            case INCLUDE:
+                return snapshotReturn;
+            case SKIP:
+                return (snapshotReturn == ReturnCode.INCLUDE) ? ReturnCode.SKIP: ReturnCode.NEXT_COL;
+            default:
+                return userRes;
+        }
+
+    }
+
+
+    private boolean isCellInSnapshot(Cell v) throws IOException {
+        if (shadowCellCache.containsKey(v.getTimestamp()) &&
+                hbaseTransaction.getStartTimestamp() >= shadowCellCache.get(v.getTimestamp())) {
+            return true;
+        }
+        if (snapshotFilter.getTSIfInTransaction(v, hbaseTransaction).isPresent()) {
+            return true;
+        }
+        Optional<Long> commitTS = snapshotFilter.getTSIfInSnapshot(v, hbaseTransaction, shadowCellCache);
+        if (commitTS.isPresent()) {
+            shadowCellCache.put(v.getTimestamp(), commitTS.get());
+            return true;
+        }
+        return false;
+    }
+
+
+    @Override
+    public void reset() throws IOException {
+        shadowCellCache.clear();
+        familyDeletionCache.clear();
+        if (userFilter != null) {
+            userFilter.reset();
+        }
+    }
+
+    @Override
+    public boolean filterRow() throws IOException {
+        if (userFilter != null) {
+            return userFilter.filterRow();
+        }
+        return super.filterRow();
+    }
+
+
+    @Override
+    public boolean filterRowKey(byte[] buffer, int offset, int length) throws IOException {
+        if (userFilter != null) {
+            return userFilter.filterRowKey(buffer, offset, length);
```

Comment:
Does this function call before each filterRowKey(Cell) ? That's what the documentation say. If it is true then the whole implementation should be different...

"Filters a row based on the row key. If this returns true, the entire row will be excluded. If false, each KeyValue in the row will be passed to filterCell(Cell) below"

Reply:
I think this ok (this is what Tephra does).